consumerBuilder.go

```diff
@@ -35,209 +31,149 @@ import (
-			sp, ok := c.dlqClusterToSaramaProducer[topic.DLQ.Cluster]
-			if !ok {
-				return errors.New("clusterConsumer builder failed to find expected sarama producer")
-			}
-			dlqMap[topic.DLQ.HashKey()] = consumer.NewDLQ(topic.DLQ.Name, topic.DLQ.Cluster, sp, c.scope, c.logger)
+		topicConsumerMap[topic.Name] = consumer.NewTopicConsumer(
+			topic.Name,
+			c.msgCh,
+			dlq,
+			c.options,
+			c.scope,
+			c.logger,
+		)
+		c.clusterTopicConsumerMap[topic.Topic.Cluster] = topicConsumerMap
+	}
+
+	// build cluster consumers
+	for cluster, topicConsumers := range c.clusterTopicConsumerMap {
+		// get topic list to consume
+		topicList := make([]string, 0, len(topicConsumers))
+		for topic := range topicConsumers {
+			topicList = append(topicList, topic)
 		}
 
-		cc, err := f(c.config.GroupName, cluster, c.opts, topicList, c.msgCh, sc, dlqMap, c.scope, c.logger)
+		// get broker list
+		brokerList, err := c.resolver.ResolveIPForCluster(cluster)
 		if err != nil {
-			c.buildErrors.AddAll(topicList, err)
-			continue
+			return nil, err
 		}
 
-		clusterConsumerMap[cluster] = cc
-		outputTopicList = append(outputTopicList, topicList...)
-	}
-
-	c.clusterToConsumer = clusterConsumerMap
-	c.topics = outputTopicList
-	return nil
-}
-
-func (c *consumerBuilder) buildSaramaProducerMap(f func([]string) (sarama.SyncProducer, error)) {
-	outputTopicList := make([]kafka.ConsumerTopic, 0, 10)
-	saramaProducerMap := make(map[string]sarama.SyncProducer)
-
-	dlqClusterTopicsMap := c.dlqClusterTopicsMap()
-	for cluster, topicList := range dlqClusterTopicsMap {
-		brokerList := topicList[0].DLQ.BrokerList
-		sp, err := f(brokerList)
+		// make new sarama consumer
+		saramaConsumer, err := c.constructors.NewSaramaConsumer(brokerList, c.config.GroupName, topicList, c.saramaClusterConfig)
 		if err != nil {
-			c.buildErrors.AddAll(topicList, err)
-			continue
+			return nil, err
 		}
-		saramaProducerMap[cluster] = sp
-		outputTopicList = append(outputTopicList, topicList...)
-	}
-
-	c.topics = outputTopicList
-	c.dlqClusterToSaramaProducer = saramaProducerMap
-}
-
-func (c *consumerBuilder) buildSaramaConsumerMap(f func([]string, string, []string, *cluster.Config) (consumer.SaramaConsumer, error)) {
-	outputTopicList := make([]kafka.ConsumerTopic, 0, 10)
-	saramaConsumerMap := make(map[string]consumer.SaramaConsumer)
 
-	clusterTopicMap := c.clusterTopicsMap()
-	for cluster, topicList := range clusterTopicMap {
-		brokerList := topicList[0].BrokerList
-		sc, err := f(brokerList, c.config.GroupName, topicList.TopicNames(), c.saramaConfig)
-		if err != nil {
-			c.buildErrors.AddAll(topicList, err)
-			continue
+		// make cluster consumer
+		c.clusterConsumerMap[cluster] = consumer.NewClusterConsumer(
+			cluster,
+			saramaConsumer,
+			topicConsumers,
+			c.scope,
+			c.logger,
+		)
+
+		// pass reference sarama consumer to topic consumers
+		for _, topicConsumer := range topicConsumers {
+			topicConsumer.SetSaramaConsumer(saramaConsumer)
```

Comment:
this set can be avoided, if you change your build the following way:
```
- go over config and build a map of cluster -> topicList (but don't construct anything yet)
- for cluster, topicList in map[cluster]topicList:
        brokers, err = resolve(cluster)
        if err != nil { closePreviouslyBuiltClusters(); return err }
        saramaConsumer = buildSaramConsumer(cluster)
        for topic in topicList:
             topicConsumer = buildTopicConsumer(saramaConsumer, topic)
             clusterToTopicConsumers[topic] = topicConsumer
        clusterConsumers[cluster] = newClusterConsumer(clusterToTopicConsumers)
```

Reply:
Done